pkg/controller/controller_message_builder_test.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+// MessageBuilder builds Event Type messages to help unit tests create these strings
+// Example usage:
+// mb := new(MessageBuilder).warning().reson("ReasonForError).msg("Error: hello world.")
+// mb.String()
+type MessageBuilder struct {
+	eventMessage  string
+	reasonMessage string
+	message       string
+}
+
+// normal sets the event type field to normal for the message builder in <type> <reason> <message>
+func (mb *MessageBuilder) normal() *MessageBuilder {
+	mb.eventMessage = corev1.EventTypeNormal
+	return mb
+}
+
+// warning sets the event type field to warning for the message builder in <type> <reason> <message>
+func (mb *MessageBuilder) warning() *MessageBuilder {
+	mb.eventMessage = corev1.EventTypeWarning
+	return mb
+}
+
+// reason sets the event reason field for the message builder in <type> <reason> <message>
+func (mb *MessageBuilder) reason(reason string) *MessageBuilder {
+	mb.reasonMessage = reason
+	return mb
+}
+
+// msg appends a message to the message builder
+func (mb *MessageBuilder) msg(msg string) *MessageBuilder {
+	space := ""
+	if mb.message > "" {
+		space = " "
+	}
+	mb.message = fmt.Sprintf(`%s%s%s`, mb.message, space, msg)
+	return mb
+}
+
+// msgf appends a message after formatting to the message builder
+func (mb *MessageBuilder) msgf(format string, a ...interface{}) *MessageBuilder {
+	msg := fmt.Sprintf(format, a...)
+	return mb.msg(msg)
+}
+
+// msgCreateServiceBindingError adds a message in the following form:
+// "Error creating ServiceBinding for ServiceInstance %q of ClusterServiceClass (K8S: %q ExternalName: %q) at ClusterServiceBroker %q:"
+func (mb *MessageBuilder) msgCreateServiceBindingError(serviceInstance, serviceClassK8S, serviceClassExternalName, broker string) *MessageBuilder {
```

Comment:
What's the decision plane for when to create one of these special-purpose methods?

Reply:
My gut is that the special-purpose methods might turn into more trouble than they're worth.

Reply:
It was used more that a few times and it was tedious and error prone to bring along the string each time. I was finding that minor copy/pasta errors were causing tests to break and making a msg* function really helped out.

I would not do this for production type code, but for tests I think it feels appropriate.